### PR TITLE
Fix Legacy & NoJit

### PR DESCRIPTION
**Pull request by Cellule**
Includes #1990 

This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/2207)

### CODE REVIEW

**Comment by Cellule** [comment]
@dotnet-bot test nojit tests please
@dotnet-bot test legacy tests please

**Comment by Cellule** [comment]
@dotnet-bot test nojit tests please
@dotnet-bot test legacy tests please

**Comment by Cellule** [comment]
The failure seems to all be slow tests.
I guess we run slow tests on legacy build.
I will merge this for now and investigate the slow tests failures

**Comment by dilijev** [comment]
Yes, it seems to be the case that slow tests are failing right now in release/1.4; https://github.com/Microsoft/ChakraCore/wiki/Build-Status-%28release-1.4%29#builds-with-slow-tests

And yes, I believe we run the slow tests in legacy config because the only automatic legacy builds are daily.

File: lib/Jsrt/ChakraCore.h [hunk 1]
```diff
@@ -212,7 +212,7 @@ CHAKRA_API
 /// </returns>
 CHAKRA_API
     JsCreateStringUtf8(
-        _In_ const uint8_t *content,
+        _In_ const unsigned char *content,
```
Comment (obastemur) [inline lib/Jsrt/ChakraCore.h]: Please do not update this file with `unsigned char`. See #1918

File: bin/ch/ChakraRtInterface.h [hunk 2]
```diff
@@ -368,8 +368,8 @@ class ChakraRTInterface
...
     static JsErrorCode WINAPI JsSerialize(JsValueRef script, ChakraBytePtr buffer, unsigned int *bufferSize, JsParseScriptAttributes parseAttributes) { return HOOK_JS_API(Serialize(script, buffer, bufferSize, parseAttributes)); }
     static JsErrorCode WINAPI JsRunSerialized(ChakraBytePtr buffer, JsSerializedLoadScriptCallback scriptLoadCallback, JsSourceContext sourceContext, JsValueRef sourceUrl, JsValueRef * result) { return HOOK_JS_API(RunSerialized(buffer, scriptLoadCallback, sourceContext, sourceUrl, result)); }
-    static JsErrorCode WINAPI JsCopyStringUtf8(JsValueRef value, uint8_t* buffer, size_t bufferSize, size_t* written) { return HOOK_JS_API(CopyStringUtf8(value, buffer, bufferSize, written)); }
-    static JsErrorCode WINAPI JsCreateStringUtf8(const uint8_t *content, size_t length, JsValueRef *value) { return HOOK_JS_API(CreateStringUtf8(content, length, value)); }
+    static JsErrorCode WINAPI JsCopyStringUtf8(JsValueRef value, unsigned char* buffer, size_t bufferSize, size_t* written) { return HOOK_JS_API(CopyStringUtf8(value, buffer, bufferSize, written)); }
```
Comment (obastemur) [inline bin/ch/ChakraRtInterface.h]: similarly, all these `unsigned char` will be updated to `char`

**Comment by Cellule** [comment]
@obastemur Would it be possible to submit like this so this PR doesn't change any functionality, but simply make compilation work.
We will be able to change `unsigned char` to `char` afterward in a more relevant PR that addresses the issue

**Comment by Cellule** [comment]
Okay so I will be submitting this PR as it is.
We should address #1918 in a separate PR

**Comment by obastemur** [comment]
> Okay so I will be submitting this PR as it is.
We should address #1918 in a separate PR

Can someone share a `TECHNICAL` reason for merging this AS IS ?

@Cellule @dilijev ?

**Comment by Cellule** [comment]
The technical reason is to avoid technical problems that could occur with such a change. 
The issue you raised can have other impacts or compatibility issues with type conversions, plus the point of this pr is to fix a build break, not implement a change in the api.

As far as api wise, nothing has changed except we don't use a typedef that can be missing in some compilers.
The issue #1918 can still be addressed by someone who has better context than me in the matter.

**Comment by obastemur** [comment]
> we don't use a typedef that can be missing in some compilers.

I can't think of a compiler (in last XX years) that is missing `typedef`. Could you give me one that we support but that doesn't have `typedef` support?

FYI; we already use `typedef` on user facing header files. (see ChakraCore.h)

 > The issue you raised can have other impacts or compatibility issues with type conversions

Sorry, I don't understand this either. How it is different than defining `unsigned short` directly ? My suggestion was `use typedef ... when char_t is not available`.. What exactly this supposed to break?

> The issue #1918 can still be addressed by someone who has better context than me in the matter.

I thought that's why we had the reviews here.

**Comment by Cellule** [comment]
Sorry, what I meant is not to support compiler that don't have the ability to use typedef, I meant these typedef `uint8_t` and `uint16_t` in particular in those files.

I'm starting to think there was some confusion here. 
I thought your initial review comment was asking me to change from `uint8_t` to `char` which is a change in the typing instead of changing `uint8_t` to `unsigned char`.
If not, I'm sorry, but I didn't understand what you were asking.

**Comment by dilijev** [comment]
This PR was intended to fix legacy (Dev12) compiler errors only. No typed or semantics were changed (the only types changed were substituted for equivalent types that fixed the legacy compiler errors).

Since that was precisely the scope of this PR, and precisely what this PR achieved, I don't see why anything else needs to be addressed here. Other issues should be discussed in the appropriate issues, and fixed in a separate PR.

**Comment by obastemur** [comment]
> No actual types or code semantics were changed (the only source-code types changed were substituted for equivalent types that fixed the legacy compiler errors, as apparently uint8_t was not supported in the legacy compiler).

@dilijev this is a very interesting approach to builtin C++11 types and the history of Unicode. Why all these new types are appeared with C++11 could be a good starting point.

IMO the change here is beyond a style update and more importantly this sort of PRs shouldn't be rushed on merging. Also they should be reviewed carefully.

We have been discussing a type change in ChakraCore.h for a while just in order to update from one type to another virtually. No rush. Better we do this discussion on Github so we can receive the valuable opinions of our community.

If we are into fixing a legacy build break, we can perfectly target <C++11 compilers as given below;

```
#if (defined(_MSC_VER) && _MSC_VER < 1900) || (__cplusplus <= 199711L) // !C++11
typedef unsigned short char16_t;
typedef unsinged int uint16_t;
#endif
``` 
I will update this file back with the planned Jsrt changes.

Thanks for the discussion.

**Comment by dilijev** [comment]
@obastemur You're right that a change which could affect API compatibility shouldn't be rushed. However there is already good issue in which to have the discussion continue rather than in an already-closed PR.

IMO this change was fine to fix a build issue (which was urgent given the impending release), with no API incompatibilities being introduced as of this change (except in APIs which are already incompatible with a previous release). It may be the case that the better approach here was to add the conditionally-compiled typedefs you mentioned, but on the other hand, that would mean that any VS 2013 project wanting to use these APIs would have to either use the compatible existing type or create the same typedefs on their side as well. I discussed this point with @Cellule offline and he decided this was the better approach to fix the build problem and unblock the test coverage for those builds.

Given that a bigger (compatibility) change to these APIs had already been decided as of #1918 and your PR implements it, I think we can just move on to worrying about the final shape of the APIs, rather than the nature of this change, and of course there are better issues and PRs to discuss this.

**Pull request merged**